src/caches/beliefpropagationcache.jl

```diff
@@ -0,0 +1,221 @@
+function default_contractor(contract_list::Vector{ITensor}; kwargs...)
+  return contract_exact(contract_list; kwargs...)
+end
+default_contractor_kwargs() = (; normalize=true, contraction_sequence_alg="optimal")
+@traitfn default_bp_niters(g::::(!IsDirected)) = is_tree(g) ? 1 : nothing
+@traitfn function default_bp_niters(g::::IsDirected)
+  return default_bp_niters(undirected_graph(underlying_graph(g)))
+end
+function message_tensor_diff(mt_a::Vector{ITensor}, mt_b::Vector{ITensor})
+  LHS, RHS = contract(mt_a), contract(mt_b)
+  return 0.5 *
+         norm((denseblocks(LHS) / sum(diag(LHS))) - (denseblocks(RHS) / sum(diag(RHS))))
+end
+
+struct BeliefPropagationCache{PTN,MTS}
+  partitioned_itensornetwork::PTN
+  message_tensors::MTS
+end
+
+#Constructors...
+function BeliefPropagationCache(ptn::PartitionedGraph; mt_storage=default_mt_storage)
+  mts = default_mt_storage(ptn)
+  return BeliefPropagationCache(ptn, mts)
+end
+
+function BeliefPropagationCache(tn::ITensorNetwork, partitioned_vertices; kwargs...)
+  ptn = PartitionedGraph(tn, partitioned_vertices)
+  return BeliefPropagationCache(ptn; kwargs...)
+end
+
+partitioned_itensornetwork(bpc::BeliefPropagationCache) = bpc.partitioned_itensornetwork
+message_tensors(bpc::BeliefPropagationCache) = bpc.message_tensors
+function tensornetwork(bpc::BeliefPropagationCache)
+  return unpartitioned_graph(partitioned_itensornetwork(bpc))
+end
+function NamedGraphs.partitioned_graph(bpc::BeliefPropagationCache)
+  return partitioned_graph(partitioned_itensornetwork(bpc))
+end
+
+function initial_message_tensor(
+  bpc::BeliefPropagationCache, edge::PartitionEdge; mt_constructor=default_mt_constructor
+)
+  ptn = partitioned_itensornetwork(bpc)
+  src_e_itn = subgraph(ptn, src(edge))
+  dst_e_itn = subgraph(ptn, dst(edge))
+  inds_e = commoninds(src_e_itn, dst_e_itn)
+  return mt_constructor(inds_e)
+end
+
+function message_tensor(
+  bpc::BeliefPropagationCache, edge::PartitionEdge; mt_constructor=default_mt_constructor
+)
+  mts = message_tensors(bpc)
+  return haskey(mts, edge) ? mts[edge] : initial_message_tensor(bpc, edge; mt_constructor)
+end
+function message_tensors(
+  bpc::BeliefPropagationCache, edges::Vector{PartitionEdge}; kwargs...
+)
+  return [message_tensor(bpc, edge; kwargs...) for edge in edges]
+end
+
+function copy(bpc::BeliefPropagationCache)
+  return BeliefPropagationCache(
+    copy(partitioned_itensornetwork(bpc)), copy(message_tensors(bpc))
+  )
+end
+
+function default_bp_niters(bpc::BeliefPropagationCache)
+  return default_bp_niters(partitioned_graph(bpc))
+end
+function default_edge_sequence(bpc::BeliefPropagationCache)
+  return default_edge_sequence(partitioned_itensornetwork(bpc))
+end
+
+"""
+Compute message tensor as product of incoming mts and local state
+"""
+function updated_message_tensor(
+  bpc::BeliefPropagationCache,
+  edge::PartitionEdge;
+  contractor=default_contractor,
+  contractor_kwargs=default_contractor_kwargs(),
```

Comment:
I'm going back and forth about the name for this. I may just prefer the name `contract` and `contract_kwargs`. In general, I think it is better to name functions that are performing an action as a verb (unless they are named after what they are outputting, in which case they would be a noun with the name of the thing they are outputting).

Reply:
Another alternative is to be more agnostic about the action it is doing, and just call it `update`. I'm worried that the code is a bit too "zoomed in" on the particular application we have in mind (simple BP and partitioned BP), while missing the bigger picture of the BP algorithm, i.e. that the messages and updates can be pretty arbitrary but the BP algorithm has a common structure. I think that considering that broader context can help us think of better names, code structures, etc.

So for example, I could imagine a function designed like this:
```julia
function update_message(
  bp_cache::BeliefPropagationCache,
  edge::PartitionEdge;
  update=default_update(bp_cache),
  update_kwargs=default_update_kwargs(bp_cache),
)
  vertex = src(edge)
  factor_vertex = factor(bp_cache, vertex) # Not sure if this is the best name, but basically it just the subgraph
  incoming_messages_edge = incoming_messages(bp_cache, edge)
  return update(factor_vertex, incoming_messages_edge; update_kwargs...)
end
```
`factor`, `incoming_message`, and `update` as an interface are agnostic about what BP is being run on, what the message are, and how the updates might be done, but we can make specialized versions of `update` and `factor` for the case of running partitioned BP on a tensor network with messages that are `Vector{ITensor}`.

Reply:
Okay I have now designed the `update_message` function this way!